### PR TITLE
[pubsub] Prioritise internal subscribers (e.g. reactor) over external (e.g. RPC)

### PR DESCRIPTION
**Pull request by melekes**
Refs #847

* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG.md

### CODE REVIEW

File: consensus/reactor.go [hunk 1]
```diff
@@ -49,6 +48,12 @@ func NewConsensusReactor(consensusState *ConsensusState, fastSync bool) *Consens
 		conS:     consensusState,
 		fastSync: fastSync,
 	}
+	// XXX: modifing state to send us new round steps, votes and proposal heartbeats
+	consensusState.reactorChs = &reactorChs{
```
Comment (melekes) [inline consensus/reactor.go]: this is not good, isn't it? should we bring `events` pkg to avoid doing this?

**Comment by melekes** [comment]
FOR REVIEW: https://github.com/tendermint/tendermint/pull/1574/files/d74efa6ac7779760e5f7209c7e2f391be9c53f4b..fffc9c707bcfd9c34b5c332584358edf37bea948

File: consensus/state.go [hunk 2]
```diff
@@ -509,6 +520,7 @@ func (cs *ConsensusState) newStep() {
 	// newStep is called by updateToStep in NewConsensusState before the eventBus is set!
 	if cs.eventBus != nil {
 		cs.eventBus.PublishEventNewRoundStep(rs)
+		cs.evsw.FireEvent(types.EventNewRoundStep, &cs.RoundState)
```
Comment (ebuchman) [inline consensus/state.go]: why not pass rs here ?

**Comment by ebuchman** [comment]
> I had to alter events package

What required this ?

**Comment by melekes** [comment]
> What required this ?

Not sure how it was done before, but the problem is in order to catch first NewRoundStep I need to subscribe to them before SwitchToConsensus emits it (https://github.com/tendermint/tendermint/pull/1574/commits/30b56886cba7d4f8d608d22e517c07408d566cd8#diff-f50d554d9c7f92cf7686b247b8c91419L85).  See https://github.com/tendermint/tendermint/pull/1574/commits/30b56886cba7d4f8d608d22e517c07408d566cd8

**Comment by ebuchman** [comment]
Sounds good. Test failure:

```
--- FAIL: TestPEXReactorRunning (10.16s)
	pex_reactor_test.go:409: expected all switches to be connected to at least one peer (switches: 0 => {outbound: 0, inbound: 1}, 1 => {outbound: 1, inbound: 1}, 2 => {outbound: 2, inbound: 0}, )
FAIL
coverage: 73.7% of statements
FAIL	github.com/tendermint/tendermint/p2p/pex	19.873s
Exited with code 1
```

I can reproduce locally on this branch but not on develop. What's that about !?

**Comment by ebuchman** [comment]
Oh the message is wrong - it actually expects everyone to be fully connected.  And now I can't reproduce it ...

**Comment by ebuchman** [comment]
Hmm I reran the test on Circle and it timedout of the consensus test.

Ran the consensus tests locally and it failed on `TestMempoolProgressAfterCreateEmptyBlocksInterval`

Looks like this might be introducing some non-determinism ?

**Comment by melekes** [comment]
Hmm.. I will take a look tmrw

**Comment by melekes** [comment]
> Looks like this might be introducing some non-determinism ?

maybe, but how? all that got changed is the way CS reactor gets its events.

**Comment by ebuchman** [comment]
Yeh especially given the test failure in p2p. Maybe the change is just helping surface previously lurking non-determinism. We're also seeing it in https://github.com/tendermint/tendermint/pull/1607

**Pull request merged**